rcldotnet/Timer.cs

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2023 Queensland University of Technology.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Runtime.InteropServices;
+using ROS2.Utils;
+
+namespace ROS2
+{
+    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+    internal delegate void TimerCallback(IntPtr timer, Duration elapsed);
+
+    internal static class TimerDelegates
+    {
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        internal delegate RCLRet NativeRCLTimerFunctionType(SafeTimerHandle timerHandle);
+
+        internal static NativeRCLTimerFunctionType native_rcl_timer_call = null;
+
+        static TimerDelegates()
+        {
+            var dllLoadUtils = DllLoadUtilsFactory.GetDllLoadUtils();
+            IntPtr nativeLibrary = dllLoadUtils.LoadLibrary("rcldotnet");
+
+            dllLoadUtils.RegisterNativeFunction(nativeLibrary, nameof(native_rcl_timer_call), out native_rcl_timer_call);
+        }
+    }
+
+    public sealed class Timer
+    {
+        private readonly Action<TimeSpan> _callback;
+
+        // The garbage collector will eventually try to clean up the delegate if nothing on the .NET side is holding on to it.
+        private readonly TimerCallback _internalCallback;
+
+        internal Timer(Clock clock, TimeSpan period, Action<TimeSpan> callback)
+        {
+            _callback = callback;
+
+            SafeTimerHandle handle = new SafeTimerHandle();
+            _internalCallback = OnTimer;
```

Comment:
It's getting late. Note for myself or for someone to answer:

Does passing the delegate to the native side work without the GC moving the delegate object and causing calls later on to fail?

Reply:
The following applies to both the callbacks in `Timer` and `Clock`.

Not storing a reference to the delegate on the .NET side does cause a failure to call the callback over the interop after garbage collection. If I was to just send `OnTimer` as in the following:

`RCLRet ret = RCLdotnetDelegates.native_rcl_create_timer_handle(ref handle, clock.Handle, new Duration(period), OnTimer);`

the resulting generated IL ends up creating a local variable which will fall out of scope and be collected with the next `GC.Collect` - the program then crashes on the next `Timer` callback.

I have tested this solution by creating a program which intentionally causes substantial `GC.Collect` - ensuring the callback continues to function.

Reply:
Not storing the reference to the delegate causes the GC to not collect the delegate, but if the delegate is alive it will be moved sooner or later to another address to compact the memory. That is the behaviour I was afraid off but did not exactly know at the time. Did some small test project to observe this: https://gist.github.com/hoffmann-stefan/7649e3793ef3563294916027336a0815

```
$ dotnet run
.......................................
Delegate with address 5368773664 once had address 5368773840
```

Maybe you haven't observed this yet as you are using the Unity dotnet runtime which doesn't have a compacting GC, or maybe the example didn't run long enough to generate enough garbage to cause the GC to compact?

Reply:
Jup, was to fast with drawing conclusions. The c-side can't call the delegate directly at it's object address, there is a `Marshal.GetFunctionPointerForDelegate` in between that generates an "Interop Stub" that can be called from c. Most certainly this will not be reallocated as it is allocated in an special `StubHeap`. This seems to be the implementation in the runtime: https://github.com/dotnet/runtime/blob/173eb1fb769edd42c856a83ce972ba0a63b425ca/src/coreclr/vm/comdelegate.cpp#L1220-L1336

I updated the code in the gist to check the returned value from `Marshal.GetFunctionPointerForDelegate`, now this seems stable. But a quick check to compare the addresses from `Marshal.GetFunctionPointerForDelegate` and the c-side would still be good to prove that there isn't something else happening here.

Reply:
Just did a test, `Marshal.GetFunctionPointerForDelegate` is used in PInvoke and the same address is passed to the native side.